File: tools/memgate.py
#!/usr/bin/env python3
"""memgate.py — pre-flight memory-pressure gate for heavy model loads.

The fleet runs multi-GB model loads on a shared, fixed-RAM box (Apple M3 Pro, 36 GB
unified; Linux DGX). Loading a 15 GB Qwen3.6-27B GGUF on top of a `llama-server`
already holding the same model in Metal unified memory is a silent OOM: the kernel
compresses/swaps, decode collapses to a crawl (0.1-0.7 tok/s), and either the new
load or a peer's process dies. This gate makes that failure mode LOUD before launch:
it reads the real available memory, names the big holders competing for it, and
either admits the load, waits for space to free, or refuses with a structured reason.

The macOS catch (why `memory_pressure`'s "Pages free" lies under a GPU load): Apple
Silicon unified memory means a Metal `-ngl 99` model is resident as WIRED pages,
counted neither free nor purgeable, but consuming the same physical pool a new CPU
load needs. So this gate reports `wired` separately and folds a conservative
available estimate = free + purgeable - safety_margin, surfacing a HIGH-WIRED flag
when a GPU model is clearly resident (wired > 40% of RAM is the heuristic).

Usage:
    python tools/memgate.py --require-gb 20                 # one-shot; exit 0 if ok, 1 if not
    python tools/memgate.py --require-gb 20 --wait 300      # poll up to 300s for space
    python tools/memgate.py --require-gb 20 --json          # structured snapshot only
    python tools/memgate.py                                  # snapshot, no requirement

Cross-platform: macOS (`vm_stat` + `sysctl hw.memsize`) and Linux (`/proc/meminfo`).
Pure stdlib. Exits: 0 = admitted/sufficient, 1 = refused/insufficient, 2 = read error.
"""
from __future__ import annotations

import argparse
import json
import platform
import subprocess
import sys
import time

# Conservative floor reserved for the OS / compositor / shell — never hand this to a model.
SAFETY_MARGIN_GB = 2.0
# Wired pages above this fraction of RAM strongly implies a resident GPU/Metal model on Apple Silicon.
HIGH_WIRED_FRACTION = 0.40
# Big-holder RSS threshold for the "who is competing" list.
HOLDER_GB = 1.0


def _run(cmd: list[str]) -> str:
    return subprocess.run(
        cmd, check=False, capture_output=True, text=True, timeout=10
    ).stdout


def read_darwin() -> dict:
    out = _run(["vm_stat"])
    page_size = int(_run(["sysctl", "-n", "hw.pagesize"]).strip() or "16384")
    total = int(_run(["sysctl", "-n", "hw.memsize"]).strip())
    vals = {}
    for line in out.splitlines():
        line = line.strip().rstrip(".")
        if ":" not in line:
            continue
        key, _, num = line.partition(":")
        key = key.strip().lower()
        num = num.strip().split()[0]
        try:
            vals[key] = int(num)
        except ValueError:
            pass
    # vm_stat reports page counts; the keys we care about differ in casing/phrasing across macOS.
    free = vals.get("pages free", vals.get("free pages", 0)) * page_size
    purgeable = vals.get("pages purgeable", 0) * page_size
    wired = vals.get("pages wired down", vals.get("wired pages", 0)) * page_size
    compressed = vals.get("pages occupied by compressor", 0) * page_size
    return {
        "total_bytes": total,
        "free_bytes": free,
        "purgeable_bytes": purgeable,
        "wired_bytes": wired,
        "compressed_bytes": compressed,
        # Conservative: reclaim the purgeable, keep the safety margin, never touch wired.
        "available_bytes": max(free + purgeable - int(SAFETY_MARGIN_GB * 1e9), 0),
    }


def read_linux() -> dict:
    vals = {}
    with open("/proc/meminfo") as f:
        for line in f:
            key, _, rest = line.partition(":")
            num = rest.strip().split()[0]
            try:
                vals[key] = int(num) * 1024  # /proc/meminfo is in kB
            except ValueError:
                pass
    total = vals.get("MemTotal", 0)
    free = vals.get("MemFree", 0)
    avail = vals.get("MemAvailable", free + vals.get("Cached", 0))
    return {
        "total_bytes": total,
        "free_bytes": free,
        "available_bytes": max(avail - int(SAFETY_MARGIN_GB * 1e9), 0),
        "purgeable_bytes": vals.get("Cached", 0),
        "wired_bytes": 0,  # not a Linux concept; slab/unreclaimable approximated elsewhere
        "compressed_bytes": 0,
    }


def read_memory() -> dict:
    sysname = platform.system().lower()
    if sysname == "darwin":
        return read_darwin()
    if sysname == "linux":
        return read_linux()
    raise RuntimeError(f"unsupported platform for memgate: {sysname}")


def big_holders() -> list[dict]:
    """Processes with RSS >= HOLDER_GB — the competing tenants the operator can free."""
    out = _run(["ps", "-axo", "pid,rss,comm"])
    holders = []
    for line in out.splitlines()[1:]:
        parts = line.strip().split(None, 2)
        if len(parts) < 3:
            continue
        try:
            pid = int(parts[0])
            rss_kb = int(parts[1])
        except ValueError:
            continue
        rss_gb = rss_kb / 1e6
        if rss_gb < HOLDER_GB:
            continue
        holders.append({"pid": pid, "rss_gb": round(rss_gb, 2), "comm": parts[2][:60]})
    holders.sort(key=lambda h: h["rss_gb"], reverse=True)
    return holders


def snapshot() -> dict:
    mem = read_memory()
    total_gb = mem["total_bytes"] / 1e9
    wired_gb = mem.get("wired_bytes", 0) / 1e9
    snap = {
        "platform": platform.system().lower(),
        "total_gb": round(total_gb, 2),
        "free_gb": round(mem["free_bytes"] / 1e9, 2),
        "available_gb": round(mem["available_bytes"] / 1e9, 2),
        "purgeable_gb": round(mem.get("purgeable_bytes", 0) / 1e9, 2),
        "wired_gb": round(wired_gb, 2),
        "compressed_gb": round(mem.get("compressed_bytes", 0) / 1e9, 2),
        "safety_margin_gb": SAFETY_MARGIN_GB,
        "high_wired": wired_gb / total_gb > HIGH_WIRED_FRACTION if total_gb else False,
        "holders": big_holders(),
    }
    snap["note"] = (
        "wired > 40% of RAM — a Metal/GPU model is likely resident in unified memory; "
        "'available' does NOT include wired, so a new CPU model load may still OOM "
        "against the GPUresident. Stop the GPU holder first."
        if snap["high_wired"]
        else "ok"
    )
    return snap


def main() -> int:
    ap = argparse.ArgumentParser(description="memory-pressure gate for heavy loads")
    ap.add_argument("--require-gb", type=float, help="GB the upcoming load needs (weights + KV + compute)")
    ap.add_argument("--wait", type=int, default=0, help="poll up to N seconds for memory to free (default: no wait)")
    ap.add_argument("--json", action="store_true", help="emit only the structured snapshot")
    ap.add_argument("--interval", type=float, default=5.0, help="poll interval seconds (with --wait)")
    args = ap.parse_args()

    if args.json and not args.require_gb:
        print(json.dumps(snapshot(), indent=2))
        return 0

    if not args.require_gb:
        snap = snapshot()
        print(json.dumps(snap, indent=2))
        return 0

    deadline = time.time() + args.wait if args.wait else 0
    while True:
        snap = snapshot()
        have = snap["available_gb"]
        ok = have >= args.require_gb and not snap["high_wired"]
        snap["require_gb"] = args.require_gb
        snap["admit"] = ok
        snap["shortfall_gb"] = round(max(args.require_gb - have, 0.0), 2)
        if ok or not args.wait or time.time() >= deadline:
            if args.json or not sys.stdout.isatty():
                print(json.dumps(snap, indent=2))
            else:
                verdict = "ADMIT" if ok else "REFUSE"
                print(f"[memgate] {verdict}: need {args.require_gb}GB, available {have}GB "
                      f"(shortfall {snap['shortfall_gb']}GB, wired {snap['wired_gb']}GB "
                      f"{'HIGH' if snap['high_wired'] else 'ok'})")
                if not ok and snap["holders"]:
                    print("[memgate] big holders to consider stopping:")
                    for h in snap["holders"][:5]:
                        print(f"           pid {h['pid']:>6}  {h['rss_gb']:>5}GB  {h['comm']}")
            return 0 if ok else 1
        time.sleep(args.interval)


if __name__ == "__main__":
    sys.exit(main())

File: internal/memgate/memgate.go
// Package memgate checks whether a heavy model load should proceed under memory pressure.
package memgate

import (
	"bufio"
	"context"
	"fmt"
	"os"
	"os/exec"
	"runtime"
	"sort"
	"strconv"
	"strings"
	"time"

	"github.com/anthony-chaudhary/fak/internal/windowgate"
)

const (
	SafetyMarginGB    = 2.0
	HighWiredFraction = 0.40
	HolderGB          = 1.0
)

type Memory struct {
	TotalBytes      int64
	FreeBytes       int64
	PurgeableBytes  int64
	WiredBytes      int64
	CompressedBytes int64
	AvailableBytes  int64
}

type Holder struct {
	PID   int     `json:"pid"`
	RSSGB float64 `json:"rss_gb"`
	Comm  string  `json:"comm"`
}

type Snapshot struct {
	Platform       string   `json:"platform"`
	TotalGB        float64  `json:"total_gb"`
	FreeGB         float64  `json:"free_gb"`
	AvailableGB    float64  `json:"available_gb"`
	PurgeableGB    float64  `json:"purgeable_gb"`
	WiredGB        float64  `json:"wired_gb"`
	CompressedGB   float64  `json:"compressed_gb"`
	SafetyMarginGB float64  `json:"safety_margin_gb"`
	HighWired      bool     `json:"high_wired"`
	Holders        []Holder `json:"holders"`
	Note           string   `json:"note"`
	RequireGB      float64  `json:"require_gb,omitempty"`
	Admit          *bool    `json:"admit,omitempty"`
	ShortfallGB    float64  `json:"shortfall_gb,omitempty"`
}

func ParseDarwin(vmStat string, pageSize, total int64) Memory {
	vals := map[string]int64{}
	for _, line := range strings.Split(vmStat, "\n") {
		line = strings.TrimSuffix(strings.TrimSpace(line), ".")
		if !strings.Contains(line, ":") {
			continue
		}
		key, rest, _ := strings.Cut(line, ":")
		fields := strings.Fields(strings.TrimSpace(rest))
		if n, ok := parseFirstIntField(fields); ok {
			vals[strings.ToLower(strings.TrimSpace(key))] = n
		}
	}
	free := firstVal(vals, "pages free", "free pages") * pageSize
	purgeable := vals["pages purgeable"] * pageSize
	wired := firstVal(vals, "pages wired down", "wired pages") * pageSize
	compressed := vals["pages occupied by compressor"] * pageSize
	// macOS parks most reclaimable RAM in inactive (evicted-but-recoverable) and
	// speculative (readahead) pages. Counting only free+purgeable made available read
	// ~0 on a healthy steady-state Mac (free 0.2G + inactive 16.6G) and refused every
	// Metal model load in the serve --gguf admission (#10595). Activity Monitor's
	// "memory used" treats inactive as available too.
	inactive := vals["pages inactive"] * pageSize
	speculative := vals["pages speculative"] * pageSize
	return Memory{
		TotalBytes:      total,
		FreeBytes:       free,
		PurgeableBytes:  purgeable,
		WiredBytes:      wired,
		CompressedBytes: compressed,
		AvailableBytes:  max64(free+inactive+speculative+purgeable-int64(SafetyMarginGB*1e9), 0),
	}
}

func ParseLinux(meminfo string) Memory {
	vals := map[string]int64{}
	sc := bufio.NewScanner(strings.NewReader(meminfo))
	for sc.Scan() {
		key, rest, ok := strings.Cut(sc.Text(), ":")
		if !ok {
			continue
		}
		fields := strings.Fields(rest)
		if n, ok := parseFirstIntField(fields); ok {
			vals[key] = n * 1024
		}
	}
	total := vals["MemTotal"]
	free := vals["MemFree"]
	avail := vals["MemAvailable"]
	if avail == 0 {
		avail = free + vals["Cached"]
	}
	return Memory{
		TotalBytes:      total,
		FreeBytes:       free,
		AvailableBytes:  max64(avail-int64(SafetyMarginGB*1e9), 0),
		PurgeableBytes:  vals["Cached"],
		WiredBytes:      0,
		CompressedBytes: 0,
	}
}

// parseFirstIntField parses the first whitespace-separated field as a base-10 int64,
// reporting ok=false (rather than an error) when fields is empty or the field doesn't
// parse — the shared ParseDarwin/ParseLinux row-skip rule: a malformed vm_stat/meminfo
// row is silently skipped, never fatal.
func parseFirstIntField(fields []string) (int64, bool) {
	if len(fields) == 0 {
		return 0, false
	}
	n, err := strconv.ParseInt(fields[0], 10, 64)
	if err != nil {
		return 0, false
	}
	return n, true
}

func ParseHolders(psText string) []Holder {
	var holders []Holder
	lines := strings.Split(psText, "\n")
	for _, line := range lines[1:] {
		parts := strings.Fields(line)
		if len(parts) < 3 {
			continue
		}
		pid, err1 := strconv.Atoi(parts[0])
		rssKB, err2 := strconv.ParseFloat(parts[1], 64)
		if err1 != nil || err2 != nil {
			continue
		}
		rssGB := rssKB / 1e6
		if rssGB < HolderGB {
			continue
		}
		comm := strings.Join(parts[2:], " ")
		if len(comm) > 60 {
			comm = comm[:60]
		}
		holders = append(holders, Holder{PID: pid, RSSGB: round2(rssGB), Comm: comm})
	}
	sort.Slice(holders, func(i, j int) bool { return holders[i].RSSGB > holders[j].RSSGB })
	return holders
}

func BuildSnapshot(platform string, mem Memory, holders []Holder) Snapshot {
	totalGB := float64(mem.TotalBytes) / 1e9
	wiredGB := float64(mem.WiredBytes) / 1e9
	highWired := false
	if totalGB > 0 {
		highWired = wiredGB/totalGB > HighWiredFraction
	}
	note := "ok"
	if highWired {
		note = "wired > 40% of RAM - a Metal/GPU model is likely resident in unified memory; 'available' does NOT include wired, so a new CPU model load may still OOM against the GPU resident. Stop the GPU holder first."
	}
	return Snapshot{
		Platform:       platform,
		TotalGB:        round2(totalGB),
		FreeGB:         round2(float64(mem.FreeBytes) / 1e9),
		AvailableGB:    round2(float64(mem.AvailableBytes) / 1e9),
		PurgeableGB:    round2(float64(mem.PurgeableBytes) / 1e9),
		WiredGB:        round2(wiredGB),
		CompressedGB:   round2(float64(mem.CompressedBytes) / 1e9),
		SafetyMarginGB: SafetyMarginGB,
		HighWired:      highWired,
		Holders:        holders,
		Note:           note,
	}
}

func Evaluate(s Snapshot, requireGB float64) Snapshot {
	ok := s.AvailableGB >= requireGB && !s.HighWired
	s.RequireGB = requireGB
	s.Admit = &ok
	if requireGB > s.AvailableGB {
		s.ShortfallGB = round2(requireGB - s.AvailableGB)
	}
	return s
}

func ReadMemory() (Memory, error) {
	switch runtime.GOOS {
	case "darwin":
		vm, err := runOut("vm_stat")
		if err != nil {
			return Memory{}, err
		}
		ps, err := runOut("sysctl", "-n", "hw.pagesize")
		if err != nil {
			return Memory{}, err
		}
		total, err := runOut("sysctl", "-n", "hw.memsize")
		if err != nil {
			return Memory{}, err
		}
		pageSize, _ := strconv.ParseInt(strings.TrimSpace(ps), 10, 64)
		totalBytes, _ := strconv.ParseInt(strings.TrimSpace(total), 10, 64)
		return ParseDarwin(vm, pageSize, totalBytes), nil
	case "linux":
		b, err := os.ReadFile("/proc/meminfo")
		if err != nil {
			return Memory{}, err
		}
		return ParseLinux(string(b)), nil
	default:
		return Memory{}, fmt.Errorf("unsupported platform for memgate: %s", runtime.GOOS)
	}
}

func BigHolders() []Holder {
	out, err := runOut("ps", "-axo", "pid,rss,comm")
	if err != nil {
		return nil
	}
	return ParseHolders(out)
}

func CurrentSnapshot() (Snapshot, error) {
	mem, err := ReadMemory()
	if err != nil {
		return Snapshot{}, err
	}
	return BuildSnapshot(runtime.GOOS, mem, BigHolders()), nil
}

func runOut(name string, args ...string) (string, error) {
	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
	defer cancel()
	cmd := exec.CommandContext(ctx, name, args...)
	windowgate.ConfigureBackgroundCommand(cmd)
	out, err := cmd.Output()
	if err != nil {
		return "", err
	}
	return string(out), nil
}

func firstVal(m map[string]int64, keys ...string) int64 {
	for _, k := range keys {
		if v, ok := m[k]; ok {
			return v
		}
	}
	return 0
}

func max64(a, b int64) int64 {
	if a > b {
		return a
	}
	return b
}

func round2(v float64) float64 {
	if v >= 0 {
		return float64(int(v*100+0.5)) / 100
	}
	return float64(int(v*100-0.5)) / 100
}
